File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/note_vm.py
"""NoteVM — leaf VM for a single note.

Capabilities (plan §3.b.5, scenario §6.2):
    ``ISelectable``, ``IClosable``, ``IDeletable[NoteVM]``, ``ISavable[NoteVM]``,
    ``IReconstructable``.

The close / delete / save callbacks are host-supplied so the leaf stays
decoupled from the container. ``close_command`` invokes ``on_close`` (the host
wires it to ``NotesViewVM.current = None`` so the form clears).
"""

from __future__ import annotations

import dataclasses
from collections.abc import Awaitable, Callable

from vmx import (
    ComponentVMOf,
    ConstructionStatus,
    IClosable,
    IDeletable,
    IReconstructable,
    ISavable,
    ISelectable,
    MessageHub,
    PropertyChangedMessage,
    RelayCommand,
    RxDispatcher,
)
from vmx.commands import ConfirmationDecoratorCommand
from vmx.commands.protocols import Command
from vmx.messages.protocols import Message
from vmx.notifications import INotificationHub, Notification, NotificationType
from vmx.services.dispatcher import Dispatcher

from notes_showcase.models.note_model import NoteModel


class NoteVM(
    ComponentVMOf[NoteModel],
    ISelectable,
    IClosable,
    IReconstructable,
):
    """Leaf VM for a single note."""

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        model: NoteModel,
        hub: MessageHub[Message],
        dispatcher: Dispatcher,
        on_close: Callable[["NoteVM"], None] | None = None,
        on_delete: Callable[["NoteVM"], None] | None = None,
        on_save: Callable[["NoteVM"], None] | None = None,
        confirm_delete: Callable[["NoteVM"], Awaitable[bool]] | None = None,
        notification_hub: INotificationHub | None = None,
    ) -> None:
        super().__init__(
            name=name,
            hint=hint,
            initial_model=model,
            modeled_hinter=lambda m: m.title,
            on_model_changed=None,
            hub=hub,
            dispatcher=dispatcher,
        )
        self._on_close = on_close
        self._on_delete = on_delete
        self._on_save = on_save
        self._confirm_delete = confirm_delete
        self._notification_hub = notification_hub

        self._close_command = (
            RelayCommand.builder().predicate(self.can_close).task(self.close).build()
        )
        self._save_command = (
            RelayCommand.builder()
            .predicate(lambda: self.can_save(self))
            .task(lambda: self.save(self))
            .build()
        )
        # Spec §5.2.8 / §6.2: when a confirm-delete delegate is wired, wrap
        # the delete in a ConfirmationDecoratorCommand. The inner command
        # invokes `_perform_delete`, which posts a "Note deleted" notification
        # (if a hub is wired) and calls the host delete callback.
        inner_delete: Command = (
            RelayCommand.builder()
            .predicate(lambda: self.can_delete(self))
            .task(lambda: self._perform_delete(self))
            .build()
        )
        self._delete_command: Command = (
            ConfirmationDecoratorCommand(
                inner_delete,
                lambda: confirm_delete(self),
            )
            if confirm_delete is not None
            else inner_delete
        )

    # ── Convenience accessors / proxies ────────────────────────────────────
    @property
    def hub(self) -> MessageHub[Message]:
        return self._hub

    @property
    def note_id(self) -> str:
        return self.model.id

    @property
    def title(self) -> str:
        return self.model.title

    @property
    def body(self) -> str:
        return self.model.body

    @property
    def starred(self) -> bool:
        return self.model.starred

    @property
    def tags(self) -> tuple[str, ...]:
        return self.model.tags

    # ── IClosable / ISavable / IDeletable ──────────────────────────────────
    def can_close(self) -> bool:
        return self._status == ConstructionStatus.CONSTRUCTED

    def close(self) -> None:
        if self._on_close is not None:
            self._on_close(self)

    def can_delete(self, item: "NoteVM") -> bool:
        return item is self and self._status == ConstructionStatus.CONSTRUCTED

    def delete(self, item: "NoteVM") -> None:
        if not self.can_delete(item):
            return
        if self._on_delete is not None:
            self._on_delete(item)

    def _perform_delete(self, item: "NoteVM") -> None:
        """Inner delete: invokes the host callback and posts the notification.

        Called by the (decorated) DeleteCommand pipeline after the confirm
        gate has succeeded. Keeps the raw :meth:`delete` capability method
        notification-free so capability dispatch stays observable-pure.
        """
        if not self.can_delete(item):
            return
        if self._on_delete is not None:
            self._on_delete(item)
        if self._notification_hub is not None:
            self._notification_hub.post(
                Notification(
                    NotificationType.NOTIFICATION,
                    f'Note deleted: “{item.title}”',
                )
            )

    def can_save(self, item: "NoteVM") -> bool:
        return item is self and self._status == ConstructionStatus.CONSTRUCTED

    def save(self, item: "NoteVM") -> None:
        if not self.can_save(item):
            return
        if self._on_save is not None:
            self._on_save(item)

    # ── Command surface ────────────────────────────────────────────────────
    @property
    def close_command(self) -> RelayCommand:
        return self._close_command

    @property
    def save_command(self) -> RelayCommand:
        return self._save_command

    @property
    def delete_command(self) -> Command:
        """Delete command. ``ConfirmationDecoratorCommand`` when ``confirm_delete``
        was wired on the builder; raw :class:`RelayCommand` otherwise.
        """
        return self._delete_command

    # ── Model setter override — emit title / starred PCMs ──────────────────
    def _set_model(self, value: NoteModel) -> None:
        old_title = self._model.title
        old_starred = self._model.starred
        super()._set_model(value)
        if old_title != value.title:
            self._hub.send(PropertyChangedMessage.create(self, self._name, "title"))
            self._raise_property_changed("title")
        if old_starred != value.starred:
            self._hub.send(PropertyChangedMessage.create(self, self._name, "starred"))
            self._raise_property_changed("starred")

    # ── Lifecycle override — dispose owned commands ────────────────────────
    def _on_dispose(self) -> None:
        self._close_command.dispose()
        self._save_command.dispose()
        # ConfirmationDecoratorCommand is not Disposable in VMx-Py; only
        # dispose the inner RelayCommand when delete_command is undecorated.
        if isinstance(self._delete_command, RelayCommand):
            self._delete_command.dispose()
        super()._on_dispose()

    # ── Builder entry-point ────────────────────────────────────────────────
    @staticmethod
    def builder() -> NoteVMBuilder:  # type: ignore[override]
        # Narrows ComponentVMOf.builder() to the showcase NoteVMBuilder.
        return NoteVMBuilder()


# Generic capability ABCs (IDeletable[T], ISavable[T]) cannot be subclassed
# alongside ComponentVMOf[M] without an MRO conflict (multiple Generic bases).
# Register them on NoteVM instead — isinstance checks still pass, which is
# what spec §14.4 capability dispatch relies on (mirrors the C# capability
# discovery via interface checks).
IDeletable.register(NoteVM)
ISavable.register(NoteVM)


@dataclasses.dataclass(frozen=True, slots=True)
class NoteVMBuilder:
    """Immutable fluent builder for :class:`NoteVM` (spec ch. 10)."""

    _name: str | None = None
    _hint: str = ""
    _model: NoteModel | None = None
    _hub: MessageHub[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _on_close: Callable[[NoteVM], None] | None = None
    _on_delete: Callable[[NoteVM], None] | None = None
    _on_save: Callable[[NoteVM], None] | None = None
    _confirm_delete: Callable[[NoteVM], Awaitable[bool]] | None = None
    _notification_hub: INotificationHub | None = None

    def name(self, value: str) -> NoteVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> NoteVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def model(self, value: NoteModel) -> NoteVMBuilder:
        return dataclasses.replace(self, _model=value)

    def services(self, hub: MessageHub[Message], dispatcher: Dispatcher) -> NoteVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def on_close(self, callback: Callable[[NoteVM], None]) -> NoteVMBuilder:
        return dataclasses.replace(self, _on_close=callback)

    def on_delete(self, callback: Callable[[NoteVM], None]) -> NoteVMBuilder:
        return dataclasses.replace(self, _on_delete=callback)

    def on_save(self, callback: Callable[[NoteVM], None]) -> NoteVMBuilder:
        return dataclasses.replace(self, _on_save=callback)

    def confirm_delete(
        self, callback: Callable[[NoteVM], Awaitable[bool]]
    ) -> NoteVMBuilder:
        """When set, :attr:`NoteVM.delete_command` is wrapped in a
        :class:`ConfirmationDecoratorCommand` calling this delegate. Typically
        wires to ``IDialogService.confirm(f"Delete '{note.title}'?")``.
        """
        return dataclasses.replace(self, _confirm_delete=callback)

    def notification_hub(self, nh: INotificationHub) -> NoteVMBuilder:
        """When set, a successful delete (post-confirm if any) publishes a
        "Note deleted" notification.
        """
        return dataclasses.replace(self, _notification_hub=nh)

    def build(self) -> NoteVM:
        if self._name is None:
            raise ValueError("name is required")
        if self._model is None:
            raise ValueError("model is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = self._dispatcher if self._dispatcher is not None else RxDispatcher.immediate()
        return NoteVM(
            name=self._name,
            hint=self._hint,
            model=self._model,
            hub=hub,
            dispatcher=dispatcher,
            on_close=self._on_close,
            on_delete=self._on_delete,
            on_save=self._on_save,
            confirm_delete=self._confirm_delete,
            notification_hub=self._notification_hub,
        )

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/NoteVM.swift
//
// NoteVM — leaf view-model for a single note.
//
// Capabilities (scenario §6.2):
//   `Selectable`, `Deselectable`, `Closable`, `Deletable`, `Savable`,
//   `Reconstructable`.
//
// `closeCommand` invokes a host-supplied close callback. This avoids
// back-references from the leaf to its container.
//
// Cross-module subclassing enabled by ADR-0066: `hub`, `dispatcher`, and
// `_notifyPropertyChanged` are `public` on the base class.
//
import Foundation
import Combine
import VMx

/// Leaf view-model for a single note.
///
/// Mirrors the C# Avalonia `NoteVM`. Holds a `NoteModel` (equality-guarded
/// setter), proxies `noteId/title/body/starred/tags`, and exposes three
/// commands: `closeCommand`, `saveCommand`, `deleteCommand`. When a
/// `confirmDelete` delegate is wired, `deleteCommand` is wrapped in a
/// `ConfirmationDecoratorCommand`.
public final class NoteVM: ComponentVMBase,
                            Selectable, Deselectable,
                            Closable, Deletable, Savable {
    // `Deletable` and `Savable` have `associatedtype Item`; Swift resolves
    // `Item = NoteVM` from the method signatures below.
    //
    // `Reconstructable` is satisfied automatically via
    // `extension ComponentVMBase: Constructable, Destructable, Reconstructable {}`.

    // ── Private state ──────────────────────────────────────────────────────

    private var _model: NoteModel
    private let _onClose: (() -> Void)?
    private let _onDelete: ((NoteVM) async throws -> Void)?
    private let _onSave: ((NoteVM) async throws -> Void)?
    private let _confirmDelete: (() async throws -> Bool)?
    private let _notificationHub: NotificationHubProtocol?

    // ── Commands (initialization placeholders, rewired after super.init) ──

    /// Inner delete relay — always an `AsyncRelayCommand`, gated on
    /// `canDelete(self)`. Wrapped by `deleteCommand` when a confirm delegate
    /// is wired.
    private var _innerDeleteCommand: AsyncRelayCommand

    /// Convenience command wrapper for `close()`.
    public private(set) var closeCommand: RelayCommand

    /// Convenience command wrapper for `save(self)`.
    public private(set) var saveCommand: AsyncRelayCommand

    /// Convenience command wrapper for `delete(self)`.
    /// Type is `any Command` because it is a plain `RelayCommand` when no
    /// confirm delegate is wired, or a `ConfirmationDecoratorCommand` when one is.
    public private(set) var deleteCommand: any Command

    // ── Public properties ──────────────────────────────────────────────────

    /// Current note model. Equality-guarded; emits `PropertyChangedMessage`s
    /// for changed derived properties (`"title"`, `"starred"`) alongside `"model"`.
    public var model: NoteModel {
        get { _model }
        set {
            guard _model != newValue else { return }
            let oldTitle = _model.title
            let oldStarred = _model.starred
            _model = newValue
            _notifyPropertyChanged("model")
            if oldTitle != newValue.title {
                _notifyPropertyChanged("title")
            }
            if oldStarred != newValue.starred {
                _notifyPropertyChanged("starred")
            }
        }
    }

    /// Note identifier (proxy on `model`).
    public var noteId: String { _model.id }
    /// Note title (proxy on `model`).
    public var title: String { _model.title }
    /// Note body (proxy on `model`).
    public var body: String { _model.body }
    /// Whether the note is starred (proxy on `model`).
    public var starred: Bool { _model.starred }
    /// Tag list (proxy on `model`).
    public var tags: [String] { _model.tags }

    // ── Selectable / Deselectable (delegated to ComponentVMBase) ──────────
    // ComponentVMBase already provides canSelect/select/canDeselect/deselect;
    // the protocol conformance declarations above satisfy the contracts.

    // ── Closable ───────────────────────────────────────────────────────────

    public func canClose() -> Bool { isConstructed }

    public func close() {
        _onClose?()
    }

    // ── Deletable (Item = NoteVM) ──────────────────────────────────────────

    public func canDelete(_ item: NoteVM) -> Bool {
        item === self && isConstructed
    }

    public func delete(_ item: NoteVM) {
        guard canDelete(item) else { return }
        _innerDeleteCommand.execute()
    }

    // ── Savable (Item = NoteVM) ────────────────────────────────────────────

    public func canSave(_ item: NoteVM) -> Bool {
        item === self && isConstructed
    }

    public func save(_ item: NoteVM) {
        guard canSave(item) else { return }
        saveCommand.execute()
    }

    // ── Internal delete implementation ─────────────────────────────────────

    private func _performDelete(_ item: NoteVM) async throws {
        // Defensive re-guard (mirrors the C# reference) so this stays correct if
        // ever invoked from a new call site beyond the already-gated commands.
        let deletedTitle = await runOnForeground { [weak self, weak item] () -> String? in
            guard let self, let item, self.canDelete(item) else { return nil }
            return item.title
        }
        guard let deletedTitle else { return }
        try await _onDelete?(item)
        if let notificationHub = _notificationHub {
            publishNotification(VMx.Notification(
                type: .notification,
                message: "Note deleted: \u{201C}\(deletedTitle)\u{201D}"
            ), to: notificationHub)
        }
    }

    private func _performSave(_ item: NoteVM) async throws {
        let admitted = await runOnForeground { [weak self, weak item] in
            guard let self, let item else { return false }
            return self.canSave(item)
        }
        guard admitted else { return }
        try await _onSave?(item)
    }

    // ── Init ───────────────────────────────────────────────────────────────

    private init(
        name: String,
        hint: String,
        model: NoteModel,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        onClose: (() -> Void)?,
        onDelete: ((NoteVM) async throws -> Void)?,
        onSave: ((NoteVM) async throws -> Void)?,
        confirmDelete: (() async throws -> Bool)?,
        notificationHub: NotificationHubProtocol?
    ) {
        _model = model
        _onClose = onClose
        _onDelete = onDelete
        _onSave = onSave
        _confirmDelete = confirmDelete
        _notificationHub = notificationHub

        // Placeholder no-op commands are required before super.init.
        let relayPlaceholder = RelayCommand(task: nil, predicate: nil, triggers: [])
        let asyncPlaceholder = AsyncRelayCommand(
            body: nil,
            predicate: nil,
            triggers: [],
            throwOnCancel: false
        )
        _innerDeleteCommand = asyncPlaceholder
        closeCommand = relayPlaceholder
        saveCommand = asyncPlaceholder
        deleteCommand = asyncPlaceholder

        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)

        // Rewire with self-capturing closures after `self` becomes valid.

        let innerDelete = AsyncRelayCommand.builder()
            .predicate({ [weak self] in
                guard let self = self else { return false }
                return self.canDelete(self)
            })
            .task({ [weak self] in
                guard let self = self else { return }
                try await self._performDelete(self)
            })
            .build()
        _innerDeleteCommand = innerDelete

        closeCommand = RelayCommand.builder()
            .predicate({ [weak self] in self?.canClose() ?? false })
            .task({ [weak self] in self?.close() })
            .build()

        saveCommand = AsyncRelayCommand.builder()
            .predicate({ [weak self] in
                guard let self = self else { return false }
                return self.canSave(self)
            })
            .task({ [weak self] in
                guard let self = self else { return }
                try await self._performSave(self)
            })
            .build()

        if let confirmDelegate = confirmDelete {
            deleteCommand = ConfirmationDecoratorCommand(innerDelete, confirm: confirmDelegate)
        } else {
            deleteCommand = innerDelete
        }
    }

    // ── Dispose ────────────────────────────────────────────────────────────

    override public func _onDispose() {
        closeCommand.dispose()
        saveCommand.dispose()
        // Dispose the decorator (if any) AND the inner command independently —
        // ConfirmationDecoratorCommand.dispose() does not cascade to its inner
        // command, so both must be disposed when confirm is wired.
        // Idempotent when deleteCommand IS innerDelete (RelayCommand.dispose is guarded).
        (deleteCommand as? ConfirmationDecoratorCommand)?.dispose()
        _innerDeleteCommand.dispose()
        super._onDispose()
    }

    // ── Builder ────────────────────────────────────────────────────────────

    /// Returns a new empty builder for `NoteVM`.
    public static func builder() -> NoteVMBuilder {
        NoteVMBuilder()
    }

    /// Immutable fluent builder for `NoteVM` (spec ch. 10).
    ///
    /// Required fields: `name(_:)`, `model(_:)`, `services(hub:dispatcher:)`.
    /// Optional fields: `onClose`, `onDelete`, `onSave`, `confirmDelete`,
    /// `notificationHub`.
    public struct NoteVMBuilder {
        private var _name: String?
        private var _hint: String = ""
        private var _model: NoteModel?
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _onClose: (() -> Void)?
        private var _onDelete: ((NoteVM) async throws -> Void)?
        private var _onSave: ((NoteVM) async throws -> Void)?
        private var _confirmDelete: (() async throws -> Bool)?
        private var _notificationHub: NotificationHubProtocol?

        fileprivate init() {}

        /// Sets the required VM name.
        public func name(_ value: String) -> NoteVMBuilder {
            var copy = self; copy._name = value; return copy
        }

        /// Sets the optional hint.
        public func hint(_ value: String) -> NoteVMBuilder {
            var copy = self; copy._hint = value; return copy
        }

        /// Sets the required note model.
        public func model(_ value: NoteModel) -> NoteVMBuilder {
            var copy = self; copy._model = value; return copy
        }

        /// Sets the required services (hub + dispatcher).
        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> NoteVMBuilder {
            var copy = self
            copy._hub = hub
            copy._dispatcher = dispatcher
            return copy
        }

        /// Sets the optional close callback (e.g. `NotesViewVM.current = nil`).
        public func onClose(_ handler: @escaping () -> Void) -> NoteVMBuilder {
            var copy = self; copy._onClose = handler; return copy
        }

        /// Sets the optional delete callback (route to repo).
        public func onDelete(_ handler: @escaping (NoteVM) async throws -> Void) -> NoteVMBuilder {
            var copy = self; copy._onDelete = handler; return copy
        }

        /// Sets the optional save callback (route to repo).
        public func onSave(_ handler: @escaping (NoteVM) async throws -> Void) -> NoteVMBuilder {
            var copy = self; copy._onSave = handler; return copy
        }

        /// Sets the optional confirm-delete delegate.
        /// When set, `deleteCommand` is wrapped in a `ConfirmationDecoratorCommand`
        /// that awaits this delegate before invoking the inner delete.
        public func confirmDelete(_ confirm: @escaping () async throws -> Bool) -> NoteVMBuilder {
            var copy = self; copy._confirmDelete = confirm; return copy
        }

        /// Sets the optional notification hub. When set, a successful delete
        /// publishes a "Note deleted: …" notification.
        public func notificationHub(_ hub: NotificationHubProtocol) -> NoteVMBuilder {
            var copy = self; copy._notificationHub = hub; return copy
        }

        /// Validates required fields and constructs a `NoteVM`.
        ///
        /// - Throws: `BuilderValidationError` if `name`, `model`, or `services` are missing.
        public func build() throws -> NoteVM {
            guard let name = _name else {
                throw BuilderValidationError(missingField: "name")
            }
            guard let model = _model else {
                throw BuilderValidationError(missingField: "model")
            }
            guard let hub = _hub else {
                throw BuilderValidationError(missingField: "hub")
            }
            guard let dispatcher = _dispatcher else {
                throw BuilderValidationError(missingField: "dispatcher")
            }
            return NoteVM(
                name: name,
                hint: _hint,
                model: model,
                hub: hub,
                dispatcher: dispatcher,
                onClose: _onClose,
                onDelete: _onDelete,
                onSave: _onSave,
                confirmDelete: _confirmDelete,
                notificationHub: _notificationHub
            )
        }
    }
}
